integration/tests/helpers/setup/mls.py

```diff
@@ -0,0 +1,247 @@
+                key_packages[c] = self.key_package_file(kp)
+        if extra_key_packages is not None:
+            for c, kp in extra_key_packages.items():
+                key_packages[c] = self.key_package_file(kp)
+
+        assert len(key_packages) > 0
+
+        msg = self.cli(cid, 
+                "member", "add", "--group", "<group-in>",
+                "--welcome-out", welcome_file.name,
+                "--group-state-out", pgs_file.name,
+                "--group-out", "<group-out>", 
+                *key_packages.values())
+        self.new_members.extend(key_packages.keys())
+
+        welcome = welcome_file.read()
+        pgs = pgs_file.read()
+
+        return MessagePackage(
+            sender=cid,
+            message=msg,
+            welcome=welcome,
+            public_group_state=pgs)
+
+    def create_application_message(self, cid, text):
+        msg = self.cli(cid, "message", "--group", "<group-in>", text)
+
+        return MessagePackage(sender=cid, message=msg)
+
+    def leave(self, user):
+        user = obj_qid(user)
+        ctx = self.ctxs[user['domain']]
+        ctx.remove_member(user, self.conversation, user).check(status=200)
+
+        # remove states, members and new_members
+        self.state = {m:s for m, s in self.state.items() if m.user != user}
+        self.members = [m for m in self.members if m.user != user]
+        self.new_members = [m for m in self.new_members if m.user != user]
+
+    def send_and_consume_message(self, mp):
+        ctx = self.ctxs[mp.sender.user['domain']]
+        events = ctx.mls_message(mp.sender.user, mp.message) \
+                    .check(status=201).json()
+        self.consume_message(mp)
+        if mp.welcome:
+            ctx.mls_welcome(mp.sender.user, mp.welcome).check(status=201)
+            self.consume_welcome(mp.welcome)
+        return events
+
+    def send_and_consume_commit(self, mp):
+        self.send_and_consume_message(mp)
+        self.epoch += 1
+
+    def consume_message(self, mp):
+        for cid in self.members:
+            if cid == mp.sender: continue
+            self.consume_message_for(cid, mp.message)
+
+    def consume_message_for(self, cid, message):
+        self.cli(cid, "consume", "--group", "<group-in>", "--group-out", "<group-out>",
+                 "--signer-key", self.removal_key(), "-", stdin=message)
+
+    def consume_welcome(self, welcome):
+        for cid in self.new_members:
+            state = self[cid]
+            assert state.group is None
+            self.cli(cid,
+                "group", "from-welcome", "--group-out", "<group-out>", "-",
+                stdin=welcome)
+
+        self.members.extend(self.new_members)
+        self.new_members = []
+
+    def cli(self, cid, *args, stdin=None):
+        state = self[cid]
+
+        subst = {}
+
+        group_in_file = None
+        if '<group-in>' in args:
+            group_in_file = tempfile.NamedTemporaryFile(mode='w+', dir=self.directory, delete=False)
```

Comment:
these are not used in a contextmanager, so won't necessarily get cleaned up?

Reply:
It will be cleaned up on destruction. That's good enough here. The file is stored in a temporary directory anyway, which is used via a context manager, so it will be deleted when that is cleaned up anyway.

Reply:
Because self.directory itself is cleaned up, and it's created in there? If that's the case, please add a commen.